### PR TITLE
Defenses & evals: Unify chat templating

### PR DESCRIPTION
**Pull request by tomtseng**
## Summary

Unify all eval prompt formatting on `tokenizer.apply_chat_template()` by making the template choice an explicit config variable (`template_name` on `ModelConfig`).

### Motivation

The codebase had a few different formatting mechanisms: (1) manual f-string formatting with `user_prefix`/`assistant_prefix`/`end_turn`, (2) `tokenizer.apply_chat_template()` using HF built-in Jinja2 templates. This has the risk of being error-prone since there a bunch of different code paths doing the same thing. This PR makes the template choice an explicit config variable and routes all formatting through `apply_chat_template()`.

Child PR #97 refactors attack training data formatting and removes the now-redundant `user_prefix`/`assistant_prefix`/`end_turn` fields from `ModelConfig`.

## Code changes

- Add `template_name` field to `ModelConfig` with four options: `native` (use tokenizer built-in template), `generic_chat`, `instruction_response`, `plain`
- Generate Jinja2 templates from the existing `TextTemplate` prefix/suffix registry via `TextTemplate.to_jinja2()`
- Centralize template configuration in `load_tokenizer()` via `configure_tokenizer_template()` so evals don't each need their own formatting logic
- Move `format_chat_prompt` and `apply_chat_template_with_fallback` from `evals/utils.py` to `whitebox/utils/models/chat_format.py` — I hit a circular import otherwise
- Remove model-specific templates (`llama3`, `qwen`, `gpt_chat`) in favor of `native` — better to just use the template given by the model tokenizer, rather than rewriting it ourselves and potentially getting it wrong

## Behavior changes

- **StrongReject**: Previously received raw (unformatted) prompts. Now prompts go through `format_chat_prompt()`. With `template: plain` this is a no-op. With other templates, prompts now get chat formatting -- this will change eval results
- **SafetyGap**: Bug fix -- the paper code does use chat templating even though our implementation said otherwise.

## TODO before merge

if any new defenses or evals landed, make sure that they also use format_chat_prompt()

Child PR: #97

### CODE REVIEW

File: src/tamperbench/whitebox/utils/models/templates.py [hunk 1]
```diff
@@ -17,13 +29,47 @@ class TextTemplate:
...
+            "{% if add_generation_prompt %}" + self.assistant_prefix + "{% endif %}"
+        )
+
 
 TEMPLATE_REGISTRY: dict[TemplateName, TextTemplate] = {
-    TemplateName.LLAMA3: TextTemplate(
-        user_prefix="<|start_header_id|>user<|end_header_id|>\n\n",
-        assistant_prefix="<|start_header_id|>assistant<|end_header_id|>\n\n",
-        end_turn="<|eot_id|>",
-    ),
```
Comment (tomtseng) [inline src/tamperbench/whitebox/utils/models/templates.py]: this was missing <|begin_of_text|> at the start of a formatted conversation. If we want to apply this kind of formatting I think we should just specify a model name and take its tokenizer.chat_template so we are sure it matches exactly

File: tests/evals/test_ifeval.py [hunk 2]
```diff
@@ -25,9 +25,10 @@ def test_ifeval_evaluation() -> None:
             model_checkpoint="meta-llama/Llama-3.1-8B-Instruct",
             out_dir=tmpdirname,
             model_config=ModelConfig(
+                template_name=TemplateName.NATIVE,
                 user_prefix="<|start_header_id|>user<|end_header_id|>\n\n",
                 assistant_prefix="<|start_header_id|>assistant<|end_header_id|>\n\n",
-                end_turn="<|eot_id|>\n",
+                end_turn="<|eot_id|>",
```
Comment (tomtseng) [inline tests/evals/test_ifeval.py:31]: actual Llama3 template does not have new-line after end_turn.

However this doesn't really matter because child PR #97 will remove this field

File: tests/evals/test_strong_reject.py [hunk 3]
```diff
@@ -23,9 +23,10 @@ def test_strong_reject_evaluation() -> None:
             model_checkpoint="EleutherAI/pythia-14m",
             out_dir=tmpdirname,
             model_config=ModelConfig(
+                template_name=TemplateName.GENERIC_CHAT,
                 user_prefix="User: ",
                 assistant_prefix="Assistant: ",
-                end_turn="\n\n",
+                end_turn="\n",
```
Comment (tomtseng) [inline tests/evals/test_strong_reject.py:34]: actual `GENERIC_CHAT` template has `end_turn="\n"`. However this `end_turn` field will get removed by child PR #97 anyway

File: src/tamperbench/whitebox/utils/models/templates.py [hunk 4]
```diff
@@ -17,13 +23,51 @@ class TextTemplate:
...
+            + "{% elif message['role'] == 'assistant' %}"
+            + self.assistant_prefix
+            + "{{ message['content'] }}"
+            + self.end_turn
+            + "{% endif %}"
+            "{% endfor %}"
+            # Strip trailing spaces from the generation prompt to avoid BPE
+            # tokenization boundary issues: a trailing space becomes a standalone
+            # token that won't appear in the actual generation (where the space
+            # merges with the first response token).
```
Comment (tomtseng) [inline src/tamperbench/whitebox/utils/models/templates.py:65]: This is needed for #97. In #97 there's a logical change where the assistant prefix is moved from the completion to the prompt. However the assistant prefix usually ends with a trailing space. If we keep it there then the `SFTTrainer`'s completion-only loss mask gets misaligned, and the model doesn't learn the first token of the response, which weakens the attacks.
